File: sdk/go/runtime/metrics.go
package runtime

// MetricsHook receives lifecycle events for observability.
// Implement this interface to integrate with Prometheus, StatsD, OpenTelemetry, etc.
// The default is NoopMetrics which has zero overhead.
type MetricsHook interface {
	OnJobReceived(jobID, topic string)
	OnJobCompleted(jobID string, durationMs int64, status string)
	OnJobFailed(jobID string, errorMsg string)
	OnHeartbeatSent(workerID string)
}

// NoopMetrics is a no-op implementation of MetricsHook.
var NoopMetrics MetricsHook = noopMetrics{}

type noopMetrics struct{}

func (noopMetrics) OnJobReceived(string, string)        {}
func (noopMetrics) OnJobCompleted(string, int64, string) {}
func (noopMetrics) OnJobFailed(string, string)           {}
func (noopMetrics) OnHeartbeatSent(string)               {}

File: sdk/python/cap/metrics.py
"""MetricsHook for CAP SDK observability.

Implement the MetricsHook protocol to integrate with Prometheus,
StatsD, OpenTelemetry, or any other metrics system.
The default is NoopMetrics which has zero overhead.
"""

import logging
from typing import Callable, Protocol


def safe_metrics_call(logger: logging.Logger, callback: Callable[[], None]) -> None:
    """Run noncritical metrics code without letting it control job delivery."""
    try:
        callback()
    except Exception as exc:
        try:
            logger.warning(
                "metrics hook failed",
                extra={"exception_type": type(exc).__name__},
            )
        except Exception:
            pass


class MetricsHook(Protocol):
    """Receives lifecycle events for observability."""

    def on_job_received(self, job_id: str, topic: str) -> None: ...
    def on_job_completed(self, job_id: str, duration_ms: int, status: str) -> None: ...
    def on_job_failed(self, job_id: str, error_msg: str) -> None: ...
    def on_heartbeat_sent(self, worker_id: str) -> None: ...


class NoopMetrics:
    """No-op metrics implementation with zero overhead."""

    def on_job_received(self, job_id: str, topic: str) -> None:
        pass

    def on_job_completed(self, job_id: str, duration_ms: int, status: str) -> None:
        pass

    def on_job_failed(self, job_id: str, error_msg: str) -> None:
        pass

    def on_heartbeat_sent(self, worker_id: str) -> None:
        pass
